### PR TITLE
refactor(plugin): new `InputPlugin` for vue

### PR DESCRIPTION
**Pull request by YvonTre**
## ⚠️TL;DR

- This pr does **NOT** affect any part of current project in any way.
- It's a base pr for better developer's review experience **ONLY**.
- More related prs follow soon today.

## What does this pr do?

- Refactor current vuex module `Input`(#306) and component `TheVideoController`'s method `getComponentName` into a brand new vue plugin: `InputPlugin`.
- Just declare the input part you want into this plugin's options, and you will find it sitting in Vuex's `InputPlugin`(or any module name you want) module already! Just like magic.

### Current Support Properties (in vuex state name)

- `mousemoveClientPosition`
- `mousemoveComponentName`
- `mousemovePhase`
- `mousedownPhase`
- `mousedownComponentName`
- `mouseupComponentName`
- `pressedKeyCodes`
- `altKeyPressedKeyCodes`
- `shiftKeyPressedKeyCodes`
- `ctrlKeyPressedKeyCodes`
- `metaKeyPressedKeyCodes`
- `wheelPhase`
- `wheelDirection`
- `wheelDirection`
- `wheelComponentName`
- `wheelDeviceType`
- `wheelDeltaX`
- `wheelDeltaY`
- `wheelDelatZ`

### Why this pr exactly?

- Currently, global input events are used in two ways:
  - UI-related change:
    - <kbd>Space</kbd> toggles `PlayButton`'s appearance.
    - <kbd>&#8592;</kbd> / <kbd>&#8594;</kbd> toggles `VolumeIndicator`'s appearance.
    - <kbd>&#8593;</kbd> / <kbd>&#8595;</kbd> toggles `TheProgressBar`'s and `TheTimeCodes`'s appearances.
  - Control-related change:
    - <kbd>&#8592;</kbd> / <kbd>&#8594;</kbd> changes the real volume.
    - <kbd>&#8593;</kbd> / <kbd>&#8595;</kbd> changes the real progress.
    - `mousemove` invoke all the components for at least 3 seconds.
- But there is a problem: all those global events, they are seperated and duplicated,
  - `document`'s `keydown` and `wheel` event listeners for controling: 
    - [src/renderer/main.js@L1295-L1362](https://github.com/chiflix/splayerx/blob/1b3df528abaa1bf51018b85add471f08eaf8f31e/src/renderer/main.js#L1295-L1362)
  - `TheVideoController`'s `mousemove`/`mousedown`/`mouseup`, `keydown`/`keyup` and `wheel` event listeners for UI:
    - [src/renderer/components/TheVideoController.vue@L6-L11](https://github.com/chiflix/splayerx/blob/1b3df528abaa1bf51018b85add471f08eaf8f31e/src/renderer/components/PlayingView/TheVideoController.vue#L6-L11)
    - [src/renderer/components/TheVideoController.vue@L276-L278](https://github.com/chiflix/splayerx/blob/1b3df528abaa1bf51018b85add471f08eaf8f31e/src/renderer/components/PlayingView/TheVideoController.vue#L276-L278)
  - ... which means that even if the same logic for ui and control, it has to be written twice and that more and more `document` or `window`'s event listeners.
- This plugin tries to solve the problem once and for all by predefining all possible global input events, and provide them through a dynamic Vuex module.
- When a component tries to control the ui or control the real data, all it has to do is reading the getters of the module and using them in any ways it wants.
- ... and all that through a simple plugin(which can be configured through options in various ways), installed once, used with ease forever.

### What does this pr feat?

- Lazy data loading: only declare the state you want to use, and other listeners will not even respond!
- Customizable module options: module name and namespaced attribute can be configured.
- Easy component recognition: just add component with a predefined type attribute, then it will magically appear at one of `xxxComponent` getters.

## To-do

- [ ] (WIP)clear of old `Input` vuex module and global event listeners
- [ ] (coming)trackpad support
- [ ] throttling
- [ ] configurable getters
- [ ] full documentation and comments
- [ ] ... maybe some tests?

### CODE REVIEW

File: src/renderer/plugins/input/helpers/destructProperties.js [hunk 1]
```diff
@@ -0,0 +1,7 @@
+export function destructProperties(object, ...properties) {
```
Comment (ipy) [inline src/renderer/plugins/input/helpers/destructProperties.js:1]: 简短的辅助方法建议合到一个文件，过于分散不利于阅读

File: src/renderer/plugins/input/helpers/destructProperties.js [hunk 2]
```diff
@@ -0,0 +1,7 @@
...
+  const result = {};
+  properties.forEach((property) => {
+    result[property] = object[property] || {};
```
Comment (ipy) [inline src/renderer/plugins/input/helpers/destructProperties.js:4]: 这个方法不建议提取到一个单独的helper里。`destructProperties`是一个很通用的名字，但是其中的行为（默认赋值`{}`）是一个很特殊的逻辑。目前调用的不多，需要的话建议用的地方单独定义。
不过不用方法写起来也只是一句话的事情：
```js
const { mouse: mouseOptions = {} } = options;
```

File: src/renderer/plugins/input/helpers/index.js [hunk 3]
```diff
@@ -0,0 +1,8 @@
+export * from './buttonsToButtonNames';
```
Comment (ipy) [inline src/renderer/plugins/input/helpers/index.js:1]: 不推荐这种将所有模块集中到一个出口的方式。目前影响不大，但是后面如果要做打包优化之类的事情会造成一些问题。需要的话可以试试`babel-plugin-transform-imports`之类的辅助工具

File: src/renderer/plugins/input/helpers/keyboardCalculator.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
+export function keydownCalculator(existedCodes, currentCode, ignoredKeys = []) {
+  existedCodes = [...existedCodes];
+  const isIgnoredKey = (
+    ignoredKeys instanceof Array &&
```
Comment (ipy) [inline src/renderer/plugins/input/helpers/keyboardCalculator.js]: 建议用 `Array.isArray`

File: src/renderer/plugins/input/helpers/namespacedNameHelper.js [hunk 5]
```diff
@@ -0,0 +1,3 @@
+export function namespacedNameHelper(namespaced, name, namespace) {
+  return `${namespaced ? namespace : ''}/${name}`;
```
Comment (ipy) [inline src/renderer/plugins/input/helpers/namespacedNameHelper.js]: 非namespaced的话前面的`/`应该也没有吧？

File: src/renderer/components/NotificationBubble.vue [hunk 6]
```diff
@@ -35,10 +35,12 @@ import { mapGetters } from 'vuex';
 import NextVideo from '@/components/PlayingView/NextVideo.vue';
 import PrivacyBubble from '@/components/PlayingView/PrivacyConfirmBubble.vue';
 import MASPrivacyBubble from '@/components/PlayingView/MASPrivacyConfirmBubble.vue';
+import { INPUT_COMPONENT_TYPE } from '@/plugins/input';
 import Icon from './BaseIconContainer.vue';
 
 export default {
   name: 'notification-bubble',
+  type: INPUT_COMPONENT_TYPE,
```
Comment (ipy) [inline src/renderer/components/NotificationBubble.vue:43]: 为什么用`type`作为这个option的名字呢？理解下来是否响应输入应该是组件的一个属性而不是类型，可能会出现具有多种属性的情况

File: src/renderer/plugins/input/helpers/wheelDetector.js [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+    }
+  }
+
+  interval = 200;
+  constructor(interval) {
+    super();
+    if (interval) this.interval = interval;
+  }
+
+  calcalate(event) {
```
Comment (ipy) [inline src/renderer/plugins/input/helpers/wheelDetector.js]: calculate

File: src/renderer/plugins/input/helpers/destructProperties.js (hunk 1 above)
Comment (YvonTre) [reply src/renderer/plugins/input/helpers/destructProperties.js:1]: - 当时新建这些文件的时候也有过这样的阅读的考虑。但是想问下，会不会产生这样的问题：如果这些函数写在`index.js`里，在写代码的时候会不会分不清是在哪里呢？因为`index.js`这种文件其实是非常常见的。
- 如果要修改，现在的修改方案是：
  - `buttonsToButtonNames.js`，---`destructProperties.js`---，`dispatch.js`和`namespacedNameHelper.js`合到`index.js`里面去。
  - `componentStore.js`，`keyboardCalculator.js`和`wheelDetector.js`仍然放在单独的文件。

File: src/renderer/plugins/input/helpers/destructProperties.js (hunk 2 above)
Comment (YvonTre) [reply src/renderer/plugins/input/helpers/destructProperties.js:4]: 好的👌

File: src/renderer/plugins/input/helpers/keyboardCalculator.js (hunk 4 above)
Comment (YvonTre) [reply src/renderer/plugins/input/helpers/keyboardCalculator.js]: 好的👌

File: src/renderer/plugins/input/helpers/namespacedNameHelper.js (hunk 5 above)
Comment (YvonTre) [reply src/renderer/plugins/input/helpers/namespacedNameHelper.js]: 好的👌

File: src/renderer/plugins/input/helpers/wheelDetector.js (hunk 7 above)
Comment (YvonTre) [reply src/renderer/plugins/input/helpers/wheelDetector.js]: 好的👌

Comment (YvonTre) [reply src/renderer/plugins/input/helpers/wheelDetector.js]: 好的👌

File: src/renderer/components/NotificationBubble.vue (hunk 6 above)
Comment (YvonTre) [reply src/renderer/components/NotificationBubble.vue:43]: 是出于这样的考量：
  - 当时设想是有三种类型的组件：
    - `BaseUIComponent`: 主要用于提供`componentName`和喂食数据的组件，比如说`PlayButton`，它就接受自己有没有被点击到。
    - `BaseContainerComponent`: 主要用于接收数据的组件，比如`TheVideoController`， `LandingView`等。这种组件的作用是从`InputPlugin`中取数据，并向子组件（一般都是`BaseUIComponent`类型传递修饰过的数据，比如说`pressedKeyCodes` -> `volumeKeydown`这种感觉。
    - `ListenerComponent`：现在的所有`listeners`都是建立在`document`上的，以后会把`listeners`建立在类型为`ListenerComponent`的组件上。
  - 基于这样的考量，所以使用`type`作为这个属性的名字。

File: src/renderer/plugins/input/helpers/index.js (hunk 3 above)
Comment (YvonTre) [reply src/renderer/plugins/input/helpers/index.js:1]: 是不是说在`import`的时候使用`babel-plugin-transform-imports`这种方式`import`?

File: src/renderer/plugins/input/helpers/destructProperties.js (hunk 1 above)
Comment (ipy) [reply src/renderer/plugins/input/helpers/destructProperties.js:1]: 代码量不大的话我觉得一个 helpers.js 可能就解决问题了。比较赞同一个文件应该职责明确内聚性强，但是文件太多太散也会造成问题，还是需要衡量取舍一下
辅助方法这种一般不会直接找吧，主要是通过定义跳转

**Review by ipy — APPROVED** [review]

**Pull request merged**